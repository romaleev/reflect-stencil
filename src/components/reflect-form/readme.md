# reflect-form



<!-- Auto Generated Below -->


## Events

| Event           | Description | Type                     |
| --------------- | ----------- | ------------------------ |
| `formSubmitted` |             | `CustomEvent<InputData>` |


## Dependencies

### Depends on

- [reflect-input](../reflect-input)

### Graph
```mermaid
graph TD;
  reflect-form --> reflect-input
  style reflect-form fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
