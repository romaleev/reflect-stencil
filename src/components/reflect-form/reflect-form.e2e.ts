import { newE2EPage } from '@stencil/core/testing'

const MockValidInput = { name: 'Roma', email: 'email@example.com', message: 'Hi' }

describe('reflect-form', () => {
  it('renders', async () => {
    const page = await newE2EPage()
    await page.setContent('<reflect-form></reflect-form>')

    const element = await page.find('reflect-form')
    expect(element).toHaveClass('hydrated')
  });

  it('renders changes to the name data', async () => {
    const page = await newE2EPage()

    await page.setContent('<reflect-form></reflect-form>')

    const inputName = await page.find('reflect-form >>> reflect-input >>> input[name="name"]')
    const inputEmail = await page.find('reflect-form >>> reflect-input >>> input[name="email"]')
    const inputMessage = await page.find('reflect-form >>> reflect-input >>> input[name="message"]')
    const buttonSubmit = await page.find('reflect-form >>> button')

    expect(await buttonSubmit.getProperty('disabled')).toBeTruthy()

    await inputName.setProperty('value', MockValidInput.name)
    await inputEmail.setProperty('value', MockValidInput.email)
    await inputMessage.setProperty('value', MockValidInput.message)

    await page.waitForChanges()

    // TODO why button is still disabled?
    // expect(await buttonSubmit.getProperty('disabled')).toBeFalsy()

    await buttonSubmit.click()

    await page.waitForChanges()

    expect(await buttonSubmit.getProperty('disabled')).toBeTruthy()
  })
})
