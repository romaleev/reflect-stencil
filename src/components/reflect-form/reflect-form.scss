$reflect-form-main-color: #FD6400;
$reflect-form-secondary-color: #FFF;
$reflect-form-background-color: lighten(#19197C, 69%);
$reflect-form-border-color: #CCC;
$reflect-form-select-color: darken($reflect-form-main-color, 10%);
$font: 1rem Arial, Helvetica, sans-serif;

:host {
  display: block;
  form {
    font: $font;
    padding: 1rem;
    background: $reflect-form-background-color;
  }

  h1 {
    background: $reflect-form-main-color;
    padding: 1.5rem 0;
    font-size: 140%;
    font-weight: 300;
    text-align: center;
    color: $reflect-form-secondary-color;
    margin: -1rem -1rem 1rem -1rem;
  }

  button[type='submit'] {
    font: $font;
    box-sizing: border-box;
    transition: all 0.30s ease-in-out;
    width: 100%;
    padding: 0.75rem;
    background: $reflect-form-main-color;
    border: 1px solid $reflect-form-border-color;
    color: $reflect-form-secondary-color;

    &:hover {
      background: $reflect-form-select-color;
      cursor: pointer;
    }
    &:focus {
      outline: none;
      box-shadow: 0 0 5px $reflect-form-select-color;
      border: 1px solid $reflect-form-select-color;
    }

    &:disabled, &:disabled:hover {
      cursor: default;
      background-color: $reflect-form-border-color;
    }
  }

  .animated {
    animation: success-animation 1s;
    animation-iteration-count: 1;
    -webkit-animation-fill-mode: both;
    animation-fill-mode: both;
  }

  @keyframes success-animation {
    from {
      -webkit-transform: scale3d(1, 1, 1);
      -ms-transform: scale3d(1, 1, 1);
      transform: scale3d(1, 1, 1);
    }
    10%, 20% {
      -webkit-transform: scale3d(.9, .9, .9) rotate3d(0, 0, 1, -3deg);
      -ms-transform: scale3d(.9, .9, .9) rotate3d(0, 0, 1, -3deg);
      transform: scale3d(.9, .9, .9) rotate3d(0, 0, 1, -3deg);
    }
    30%, 50%, 70%, 90% {
      -webkit-transform: scale3d(1.1, 1.1, 1.1) rotate3d(0, 0, 1, 3deg);
      -ms-transform: scale3d(1.1, 1.1, 1.1) rotate3d(0, 0, 1, 3deg);
      transform: scale3d(1.1, 1.1, 1.1) rotate3d(0, 0, 1, 3deg);
    }
    40%, 60%, 80% {
      -webkit-transform: scale3d(1.1, 1.1, 1.1) rotate3d(0, 0, 1, -3deg);
      -ms-transform: scale3d(1.1, 1.1, 1.1) rotate3d(0, 0, 1, -3deg);
      transform: scale3d(1.1, 1.1, 1.1) rotate3d(0, 0, 1, -3deg);
    }
    to {
      -webkit-transform: scale3d(1, 1, 1);
      -ms-transform: scale3d(1, 1, 1);
      transform: scale3d(1, 1, 1);
    }
  }
}
