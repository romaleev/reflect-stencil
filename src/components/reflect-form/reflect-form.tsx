import { Component, Element, State, Host, h, Listen, Event, EventEmitter } from '@stencil/core'
import { validate, InputData } from '../../utils/utils'

@Component({
  tag: 'reflect-form',
  styleUrl: 'reflect-form.scss',
  shadow: true
})

/**
 * ReflectForm contains ReflectInput which contains input fields
 *
 * Submittable only if all input fields are filled and valid
 *
 * Listens for `inputChanged` event of ReflectInput component
 *
 * Emits `formSubmitted` event per successful form submission
 *
 * ```
 * <reflect-form/>
 * ```
 */
export class ReflectForm {

  @Element() private element: HTMLElement

  @State() valid = false

  inputData: InputData
  reflectForm: HTMLElement
  reflectInput: HTMLReflectInputElement

  /**
   * Listens for `inputChanged` event of ReflectInput component
   */
  @Listen('inputChanged')
  inputChangedHandler(event: CustomEvent<InputData>) {
    this.inputData = event.detail
    this.valid = validate(this.inputData)
  }

  /**
   * @event `formSubmitted` emitted per successful form submission
   * @public
   */
  @Event({
    eventName: 'formSubmitted',
    composed: true,
    cancelable: true,
    bubbles: true,
  }) onSubmit: EventEmitter<InputData>

  componentDidLoad() {
    this.reflectForm = this.element.shadowRoot.querySelector('form')
    this.reflectInput = this.element.shadowRoot.querySelector('reflect-input')
  }

  async handleSubmit(event: UIEvent) { // submit button callback
    event.preventDefault()
    if(this.valid) {
      if(this.onSubmit) {
        this.onSubmit.emit(this.inputData)
      }
      if(this.reflectForm) {
        this.reflectForm.classList.add('animated')
        this.reflectForm.addEventListener('animationend', () =>
          this.reflectForm.classList.remove('animated')) // instead of setTimeout(() => this.reflectForm.classList.remove('animated'), 1000)
      }
      this.reflectInput && await this.reflectInput.clear()
    }
  }

  render() {
    return (
      <Host>
        <form onSubmit={(event: UIEvent) => this.handleSubmit(event)} >
          <h1>CONTACT US</h1>
          <reflect-input />
          {/*<slot /> <!-- we can also put reflect-input here, it would be outside current shadowRoot -->*/}
          <button tabindex="0"
                  disabled={!this.valid}
                  type="submit">
            SUBMIT
          </button>
        </form>
      </Host>
    );
  }

}
