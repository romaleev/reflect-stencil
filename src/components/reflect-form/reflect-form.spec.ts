import { ReflectForm } from './reflect-form';
import { newSpecPage } from '@stencil/core/testing'

describe('reflect-form', () => {
  it('builds and renders', () => {
    const reflectForm = new ReflectForm()

    expect(reflectForm).toBeTruthy()
    expect(reflectForm).toMatchSnapshot();
  })

  it('should not be submittable with invalid input', async () => {
    const page = await newSpecPage({
      components: [ ReflectForm ],
      html: `<reflect-form></reflect-form>`
    })
    const reflectForm = await document.querySelector('reflect-form')
    const reflectFormButton = await reflectForm.shadowRoot.querySelector('button')
    const eventSpy = jest.fn()
    page.doc.addEventListener('formSubmitted', eventSpy)

    await reflectFormButton.click()

    expect(eventSpy).not.toHaveBeenCalled()
  })
})
