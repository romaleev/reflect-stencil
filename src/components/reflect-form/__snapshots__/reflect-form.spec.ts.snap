// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reflect-form builds and renders 1`] = `
ReflectForm {
  "onSubmit": Object {
    "emit": [Function],
  },
  "valid": false,
}
`;
